RecoPixelVertexing/PixelVertexFinding/src/PixelVertexHeterogeneousProducer.cc

```diff
@@ -0,0 +1,242 @@
+
+  // on gpu beamspot already subtracted at hit level...
+  math::XYZPoint bs(0.,0.,0.);
+  edm::Handle<reco::BeamSpot> bsHandle;
+  e.getByToken(token_BeamSpot,bsHandle);
+
+  if (bsHandle.isValid()) bs = math::XYZPoint(bsHandle->x0(),bsHandle->y0(), bsHandle->z0() ); 
+
+  // Second, make a collection of pointers to the tracks we want for the vertex finder
+  // fill z,ez
+  std::vector<float> z,ez;
+  assert(m_trks.empty());
+  for (unsigned int i=0; i<tracks.size(); i++) {
+    if (tracks[i].pt() < m_ptMin) continue;    
+    m_trks.push_back( reco::TrackRef(trackCollection, i) );
+    z.push_back(tracks[i].dz(bs));
+    ez.push_back(tracks[i].dzError());ez.back()*=ez.back();
+  }
+  if (verbose_) edm::LogInfo("PixelVertexHeterogeneousProducer") << ": Selected " << m_trks.size() << " of these tracks for vertexing\n";
+  
+  // Third, ship these tracks off to be vertexed
+  m_gpuAlgo.produce(cudaStream.id(),z.data(),ez.data(),z.size());
+
+}
+
+void PixelVertexHeterogeneousProducer::produceGPUCuda(
+    edm::HeterogeneousEvent & e, const edm::EventSetup & es,
+    cuda::stream_t<> &cudaStream) {
+
+
+  auto const & gpuProduct = m_gpuAlgo.fillResults(cudaStream.id());
+
+  
+  edm::Handle<reco::BeamSpot> bsHandle;
+  e.getByToken(token_BeamSpot,bsHandle);
+ 
+  auto vertexes = std::make_unique<reco::VertexCollection>(); 
+
+
+  float x0=0,y0=0,z0=0,dxdz=0,dydz=0;
+  std::vector<int> itrk;
+  if(!bsHandle.isValid()) {
+     edm::LogWarning("PixelVertexHeterogeneousProducer") << "No beamspot found. Using returning vertexes with (0,0,Z) ";
+  } else {
+    const reco::BeamSpot & bs = *bsHandle;
+    x0=bs.x0();y0=bs.y0();z0=bs.z0(); dxdz=bs.dxdz();dydz=bs.dydz();
+  }
+  for (unsigned int i=0; i<gpuProduct.nVertices; ++i) {
+    assert(itrk.empty());
+    auto z= gpuProduct.z[i];
+    auto x= x0 + dxdz*z;
+    auto y= y0 + dydz*z;
+    z +=z0;
+    reco::Vertex::Error err;
+    err(2,2) = 1.f/gpuProduct.zerr[i];
+    //Copy also the tracks (no intention to be efficient....)
+    for (auto k=0U; k<m_trks.size(); ++k) {
+      if (gpuProduct.ivtx[k]==int(i)) itrk.push_back(k);
+    }
+    auto nt = itrk.size();
+    assert(nt>0);
+    (*vertexes).emplace_back(reco::Vertex::Point(x,y,z), err, gpuProduct.chi2[i], nt-1, nt );
+    auto & v = (*vertexes).back();
+    for (auto k: itrk) v.add(reco::TrackBaseRef(m_trks[k]));
+    itrk.clear();
+  }
+
+  if (verbose_) {
+    edm::LogInfo("PixelVertexHeterogeneousProducer") << ": Found " << vertexes->size() << " vertexes\n";
+    for (unsigned int i=0; i<vertexes->size(); ++i) {
+      edm::LogInfo("PixelVertexHeterogeneousProducer") << "Vertex number " << i << " has " << (*vertexes)[i].tracksSize() << " tracks with a position of " << (*vertexes)[i].z() << " +- " << std::sqrt( (*vertexes)[i].covariance(2,2) );
+    }
+    
+    std::cout << ": Found " << vertexes->size() << " vertexes\n";
```

Comment:
`edm::LogInfo` ?

Reply:
it is just above, I have no clue how to make it working....